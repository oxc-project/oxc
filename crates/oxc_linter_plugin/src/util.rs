use std::path::{Component, Path, PathBuf};

use path_calculate::Calculate;

/// Makes relative path part vec with a path relative to the current directory.
///
/// ie: if current directory = /foo/
///    and path = /foo/bar/baz.txt
///   then relative path parts = [Some("bar"), Some("baz.txt")]
///
/// # Panics
/// Panics if the path component has any non-normal component after
/// a single normal path component.
pub fn make_relative_path_parts(path: &PathBuf) -> Vec<Option<String>> {
    path.related_to(Path::new("."))
        .expect("to be able to get the relative path parts")
        .components()
        .skip_while(|x| !matches!(x, Component::Normal(_)))
        .map(|path_component| {
            let Component::Normal(component) = path_component else {
                unreachable!("there should only be normal path components here")
            };
            component.to_str().map(ToOwned::to_owned)
        })
        .collect::<Vec<_>>()
}
