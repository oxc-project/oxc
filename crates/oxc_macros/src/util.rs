use syn::{parse::ParseStream, Result, Token};

/// Checks if `cond` is `true`, returning [`Err(syn::Error)`] with `msg` if it's not.
/// ## Example
/// ```ignore
/// use syn::{parse::Parse, LitStr};
///
/// struct Foo(LitStr);
///
/// impl Parse for Foo {
///     fn parse(input: ParseStream) -> Result<Self> {
///         let s = input.parse::<LitStr>()?;
///         parse_assert!(s.value() == "foo", s, "Expected 'foo'");
///         Ok(Foo(s))
///     }
/// }
/// ```
macro_rules! parse_assert {
    ($cond:expr, $toks:expr, $msg:expr) => {
        if !($cond) {
            return Err(syn::Error::new_spanned($toks, $msg));
        }
    };
}
pub(crate) use parse_assert;

/// Consume a comma token if it's present, noop otherwise
#[allow(clippy::trivially_copy_pass_by_ref)]
pub(crate) fn eat_comma(input: &ParseStream) -> Result<()> {
    if input.peek(Token!(,)) {
        input.parse::<Token!(,)>()?;
    }
    Ok(())
}
