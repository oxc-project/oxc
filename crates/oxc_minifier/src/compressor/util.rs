use oxc_ast::ast::Expression;

pub(super) fn is_console(expr: &Expression<'_>) -> bool {
    // let Statement::ExpressionStatement(expr) = stmt else { return false };
    let Expression::CallExpression(call_expr) = &expr else { return false };
    let Expression::MemberExpression(member_expr) = &call_expr.callee else { return false };
    let obj = member_expr.object();
    let Some(ident) = obj.get_identifier_reference() else { return false };
    ident.name == "console"
}
