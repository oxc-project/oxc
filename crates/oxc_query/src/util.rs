use std::hash::{Hash, Hasher};

use oxc_ast::ast::{
    Expression, JSXAttribute, JSXAttributeName, JSXAttributeValue, JSXElementName, JSXIdentifier,
    JSXMemberExpression, JSXNamespacedName, TSAccessibility, TemplateLiteral,
};
use rustc_hash::FxHasher;

use crate::{
    vertex::{VarRefVertex, Vertex},
    Adapter,
};

pub fn jsx_attribute_to_constant_string<'a>(attr: &'a JSXAttribute<'a>) -> Option<String> {
    attr.value.as_ref().and_then(|attr_value| match attr_value {
        JSXAttributeValue::String(slit) => slit.value.to_string().into(),
        JSXAttributeValue::ExpressionContainer(expr) => match &expr.expression {
            oxc_ast::ast::JSXExpression::Expression(expr) => expr_to_maybe_const_string(expr),
            oxc_ast::ast::JSXExpression::EmptyExpression(_) => None,
        },
        JSXAttributeValue::Element(_) | JSXAttributeValue::Fragment(_) => None,
    })
}

pub fn expr_to_maybe_const_string<'a>(expr: &'a Expression<'a>) -> Option<String> {
    match expr {
        Expression::StringLiteral(slit) => Some(slit.value.to_string()),
        Expression::TemplateLiteral(tlit) => {
            try_get_constant_string_field_value_from_template_lit(tlit.0)
        }
        _ => None,
    }
}

pub fn try_get_constant_string_field_value_from_template_lit(
    tlit: &TemplateLiteral,
) -> Option<String> {
    if tlit.expressions.len() == 0 && tlit.quasis.len() == 1 {
        let quasi = &tlit.quasis[0].value;
        Some(quasi.cooked.as_ref().unwrap_or(&quasi.raw).to_string())
    } else {
        None
    }
}

pub fn accessibility_to_string(accessibility: TSAccessibility) -> String {
    match accessibility {
        TSAccessibility::Private => "private",
        TSAccessibility::Protected => "protected",
        TSAccessibility::Public => "public",
    }
    .to_owned()
}

pub fn jsx_identifier_to_string(jsx_identifier: &JSXIdentifier) -> String {
    jsx_identifier.name.to_string()
}

pub fn jsx_namespaced_name_to_string(jsx_namespaced_name: &JSXNamespacedName) -> String {
    format!("{}:{}", &jsx_namespaced_name.namespace.name, &jsx_namespaced_name.property.name)
}

pub fn jsx_membexpr_to_string(jsx_member_expr: &JSXMemberExpression) -> String {
    let mut parts = vec![jsx_member_expr.property.name.to_string()];
    let mut obj = &jsx_member_expr.object;
    loop {
        match obj {
            oxc_ast::ast::JSXMemberExpressionObject::Identifier(ident) => {
                parts.push(ident.name.to_string());
                break;
            }
            oxc_ast::ast::JSXMemberExpressionObject::MemberExpression(inner_memexpr) => {
                parts.push(inner_memexpr.property.name.to_string());
                obj = &inner_memexpr.object;
            }
        }
    }
    parts.reverse();
    parts.join(".")
}

pub fn jsx_element_name_to_string(jsx_element_name: &JSXElementName) -> String {
    match jsx_element_name {
        JSXElementName::Identifier(ident) => jsx_identifier_to_string(ident),
        JSXElementName::NamespacedName(nsn) => jsx_namespaced_name_to_string(nsn),
        JSXElementName::MemberExpression(memexpr) => jsx_membexpr_to_string(memexpr),
    }
}

pub fn jsx_attribute_name_to_string(jsx_attribute_name: &JSXAttributeName) -> String {
    match jsx_attribute_name {
        JSXAttributeName::Identifier(ident) => jsx_identifier_to_string(ident),
        JSXAttributeName::NamespacedName(nsn) => jsx_namespaced_name_to_string(nsn),
    }
}

pub fn strip_parens_from_expr<'a, 'b: 'a>(
    to_strip: &'b Expression<'a>,
    strip_all: bool,
) -> &'b Expression<'a> {
    match to_strip {
        Expression::ParenthesizedExpression(paren_expr) if strip_all => {
            strip_parens_from_expr(&paren_expr.expression, true)
        }
        Expression::ParenthesizedExpression(paren_expr) => &paren_expr.expression,
        _ => to_strip,
    }
}

pub fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut hasher = FxHasher::default();
    t.hash(&mut hasher);
    hasher.finish()
}

pub fn declaration_of_varref<'b, 'c: 'b>(
    data: &VarRefVertex<'c>,
    adapter: &'b Adapter<'c>,
) -> Option<Vertex<'c>> {
    let reference_id = data.identifier_reference.reference_id.get()?;
    let reference = adapter.semantic.symbols().references.get(reference_id)?;
    let symbol_id = reference.symbol_id()?;
    Some((*adapter.semantic.symbol_declaration(symbol_id)).into())
}
