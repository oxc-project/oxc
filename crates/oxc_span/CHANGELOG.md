# Changelog

All notable changes to this crate will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does not adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) until v1.0.0.

## [0.10.0] - 2024-03-14

### Features

- Merge features `serde` and `wasm` to `serialize` (#2716)
- `impl<'a> PartialEq<str> for Atom<'a>` (#2649)
- Miette v7 (#2465)
- Remove `From<String>` and `From<Cow>` API because they create memory leak (#2628)

### Refactor

- Derive `SerAttrs` on all AST types (#2698)
- Change shape of `Language` (#2680)
- Reduce `cfg_attr` boilerplate with `SerAttrs` derive (#2669)
- Import `Tsify` to shorten code (#2665)
- "wasm" feature enable "serde" feature (#2639)
- Shorten manual TS defs (#2638)
- Simplify `Atom` (#2630)
- Remove `Atom::Compact` variant (#2629)
- Disallow struct expression constructor for `Span` (#2625)
- Make `CompactStr` immutable (#2620)
- Rename `CompactString` to `CompactStr` (#2619)

## [0.9.0] - 2024-03-05

### Refactor

- Remove `AtomImpl` (#2525)
- Move base54 method to mangler (#2523)
- Replace InlinableString with CompactString for `Atom` (#2517)

## [0.7.0] - 2024-02-09

### Features

- Fix memory leak by implementing inlineable string for oxc_allocator (#2294)

## [0.6.0] - 2024-02-03

### Features

- TypeScript definition for wasm target (#2158)

## [0.4.0] - 2023-12-08

### Refactor

- Move to workspace lint table (#1444)

## [0.3.0] - 2023-11-06

### Features

- Support react/no-render-return-value (#1042)

### Refactor

- Add an empty SPAN utility for creating AST nodes (#1067)

### Testing

- Add scoping test cases (#954)

## [0.2.0] - 2023-09-14

### Features

- Constant addition expression folding (#882)

### Refactor

- Improve code coverage in various places (#721)

