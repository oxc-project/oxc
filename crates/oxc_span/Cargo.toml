[package]
name                   = "oxc_span"
version                = "0.12.3"
publish                = true
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
miette      = { workspace = true }
compact_str = { workspace = true }

tsify        = { workspace = true, optional = true }
wasm-bindgen = { workspace = true, optional = true }
serde        = { workspace = true, features = ["derive"], optional = true }

[features]
default   = []
serialize = ["dep:serde", "dep:tsify", "dep:wasm-bindgen", "compact_str/serde"]
