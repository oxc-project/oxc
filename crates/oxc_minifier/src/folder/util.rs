use std::cmp::Ordering;

use num_bigint::BigInt;

use super::tri::Tri;

use crate::compressor::ast_util::{is_exact_int64, NumberValue};

/// ported from [closure compiler](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/PeepholeFoldConstants.java#L1250)
#[allow(clippy::cast_possible_truncation)]
pub fn bigint_less_than_number(
    bigint_value: &BigInt,
    number_value: &NumberValue,
    invert: Tri,
    will_negative: bool,
) -> Tri {
    // if invert is false, then the number is on the right in tryAbstractRelationalComparison
    // if it's true, then the number is on the left
    match number_value {
        NumberValue::NaN => Tri::for_boolean(will_negative),
        NumberValue::PositiveInfinity => Tri::True.xor(invert),
        NumberValue::NegativeInfinity => Tri::False.xor(invert),
        NumberValue::Number(num) => {
            if let Some(Ordering::Equal | Ordering::Greater) =
                num.abs().partial_cmp(&2_f64.powi(53))
            {
                Tri::Unknown
            } else {
                let number_as_bigint = BigInt::from(*num as i64);

                match bigint_value.cmp(&number_as_bigint) {
                    Ordering::Less => Tri::True.xor(invert),
                    Ordering::Greater => Tri::False.xor(invert),
                    Ordering::Equal => {
                        if is_exact_int64(*num) {
                            Tri::False
                        } else {
                            Tri::for_boolean(num.is_sign_positive()).xor(invert)
                        }
                    }
                }
            }
        }
    }
}
