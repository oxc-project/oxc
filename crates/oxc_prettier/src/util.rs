use oxc_span::Span;

pub fn is_next_line_empty(source_text: &str, span: Span) -> bool {
    source_text.chars().nth((span.end as usize) + 1).is_some_and(|c| c == '\n')
}
