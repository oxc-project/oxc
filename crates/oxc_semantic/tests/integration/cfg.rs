use std::fs;

use oxc_span::SourceType;

use crate::util::SemanticTester;

#[test]
fn test_cfg_files() {
    insta::glob!("cfg_fixtures/*.js", |path| {
        let code = fs::read_to_string(path).unwrap();
        let output = SemanticTester::new(&code, SourceType::from_path(path).unwrap());
        insta::assert_snapshot!(output.basic_blocks_printed());
        insta::assert_snapshot!(output.cfg_dot_diagram());
    });
}

// todo: private identifier from class function name should not be in cfg

// #[test]
// fn test_cfg_expected_basic_block_count() {
//     for test in [
//         ("(a => {}).length", 2),
//         ("typeof (() => {});", 2),
//         ("export default (() => {})();", 2),
//         ("(() => {})()``;", 2),
//         ("(() => {})``;", 2),
//         ("new (() => {});", 2),
//         (
//             "
//         // 1 - global scope
//         if (
//             (() => {}) // 2 - inside arrow function
//             ? 1 // 3 - consequent of conditional expression
//             : 0 // 4 - alternate
//             // 5 - after arrow expression
//         ) {
//             // 6 - inside consequent of if statement
//         }
//         // 7 - after if statement
//         ",
//             7,
//         ),
//         ("let f = () => ({}())", 2),
//         ("let a = () => ({} instanceof a);", 2),
//         ("a = () => ({} && a);", 5),
//         ("a = () => ({}() && a);", 5),
//         ("a = () => ({} && a && b);", 8),
//         ("a = () => ({} + a);", 2),
//         ("a = () => ({}()() && a);", 5),
//         ("a = () => ({}.b && a);", 5),
//         ("a = () => ({}[b] && a);", 5),
//         (" a = () => ({}`` && a);", 5),
//         ("a = () => ({} = 0);", 2),
//         ("a = () => ({}, a);", 2),
//         ("a => a instanceof {};", 2),
//         ("a => ({}().b && 0)", 5),
//         ("a => ({}().c = 0)", 2),
//         ("x => ({}()())", 2),
//         ("x => ({}()``)", 2),
//         ("x => ({}().b);", 2),
//         ("a = b => c;", 2),
//         ("x => (y = z);", 2),
//         ("x => (y += z);", 2),
//         ("f(a => ({})) + 1;", 2),
//         ("(a => ({})) || 0;", 5),
//         ("a = b => c;", 2),
//         (
//             "a = b => {
//                 return c
//             };",
//             3,
//         ),
//     ] {
//         let count = SemanticTester::js(test.0).basic_blocks_count();

//         assert_eq!(
//             test.1, count,
//             "The code: \n{}\nshould have had {} basic blocks\nhowever actually had {} basic blocks.",
//             test.0, test.1, count
//         );
//     }
// }
