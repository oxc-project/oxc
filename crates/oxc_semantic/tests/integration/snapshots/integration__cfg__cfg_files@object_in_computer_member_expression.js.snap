---
source: crates/oxc_semantic/tests/integration/cfg.rs
expression: output.basic_blocks_printed()
input_file: crates/oxc_semantic/tests/integration/cfg_fixtures/object_in_computer_member_expression.js
---
bb0: {
	statement
}
